ADR/0006-log-conventions.md

```diff
@@ -0,0 +1,112 @@
+# 6. Log Conventions
+
+Date started: 2022-November-11
+Date revised: 2023-March-8
+
+## Status
+
+Accepted
+
+## Context
+
+We need log conventions to make our offering easier to operate and maintain. These will also make it possible for us to create queries that cross components in our logs, which supports [STONE-201](https://issues.redhat.com/browse/STONE-201).
+
+## Decision: Log conventions
+
+In our controller logs, we will use structured log messages, partially formatted in JSON with key-value pairs as described below.
+
+```
+TIMESTAMP LOG_LEVEL LOGGER MESSAGE ADDITIONAL_JSON_DATA
+```
+
+For example:
+
+```
+2023-03-07T11:32:29.167Z    INFO    controllers.Component   updating devfile component name kubernetes-deploy ...   {"resource": "samburrai-tenant/go-net-http-hello-prqx", "kind": "Component"}
+2023-03-07T11:32:29.020Z    INFO    controllers.Component   API Resource changed: UPDATE {"namespace": "samburrai-tenant", "audit": "true", "resource": "go-net-http-hello-prqx", "kind": "Component", "action": "UPDATE"}
+```
+
+The meaning of **timestamp** and log level should be self evident. The **logger** name helps your team understand which part of your controller is emitting the log (usually `Log.WithName` from `zap`). The **message** is a human readable string describing the event being logged. The **json data** contains additional fields useful for searching.
+
+### 1. When did it happen?
+
+Encode timestamps in UTC/ISO-8601 format.
+
+### 2. What happened?
+
+Use the key `action` with possible values `VIEW, ADD, UPDATE, DELETE`, if applicable.
+
+### 3. Where did it happen?
+
+There is no need to identify which stonesoup subsystem the log is coming from (i.e., HAS, SPI, or
+GitOps).
+
+Consider: if an engineer is looking at logs directly in the namespace where a controller is
+deployed, then you know which service you are looking at. If an engineer is looking at logs
+centralized in cloudwatch or splunk, then the namespace from which the log came will be included
+automatically as `namespace_name`, which is sufficient to determine what stonesoup subsystem
+produced the log. See section on automatically added logs below.
+
+### 4. Who was involved?
+
+Use the key `namespace` when logging the namespace of the *targeted resource* that is being modified
+or interacted with. (Note that the `namespace_name` key is automatically added by the fluent
+collectors and reflects the namespace in which the controller is running. See section on
+automatically added logs below.)
+
+Include a `user_id` if one exists and is applicable for the event being logged.
```

Comment:
Is this ADR specific to audit logs or all logs in general?  We should avoid collecting PII such as `user_id` in regular logs but they are acceptable in audit logs.  I believe audit logs are the only ones we are forwarding to CloudWatch/Splunk so we should make this distinction clear.

Reply:
Related to this, should data like `namespace_name` that is auto-added by fluentd be considered PII as well and handled specially?

Reply:
> Is this ADR specific to audit logs or all logs in general?

I understand it to be for controller logs in general, including audit logs, in order to support the use cases in [STONE-200](https://issues.redhat.com/browse/STONE-200), [STONE-201](https://issues.redhat.com/browse/STONE-201), and [STONE-241](https://issues.redhat.com/browse/STONE-241).

> We should avoid collecting PII such as `user_id` in regular logs but they are acceptable in audit logs.

What you write makes sense.

> I believe audit logs are the only ones we are forwarding to CloudWatch/Splunk so we should make this distinction clear.

My understanding is that we desire _all_ logs to be forwarded to cloudwatch (for the purposes of our own team and SRE to debug and understand our system in production), but we desire that only audit logs be forwarded to splunk (for privileged query by security).

I am not aware of a method to easily log twice, once with a user id and once without, and ensure that one log is sent to cloudwatch (collected from pod stdout/err by fluent) while the other log is sent to splunk (forwarded how?).

Due to lack of clarity around the above, I think we should revise this to limit the scope of this ADR to being strictly about general logs, and drop this line about including a `user_id` if applicable. A consequence should be added at the end of the doc indicating that we'll need to revisit this if/when we need to include user ids specifically for audit logs in the future.

@kim-tsao, wdyt?

Reply:
@ralphbean that sounds reasonable, thanks!  Can you also clarify the question @arewm raised about `namespace_name` - can that be PII?

Reply:
I bet it will be simpler to just align in both prod and stage - make them identical. But, let's leave that decision to the Core team (unless they ask for our input).

Reply:
Updated in 0d130033b1b1bdfc4770c75ee2ebecbf23163138.

Reply:
The good news is that in our PIA, we said we would have PII in the logs. But limiting the usage of PII to "just enough" to serve our legitimate needs is the way to go.